error: no matching function for call to ‘v8::StringObject::New(v8::Local<v8::String>&)’

yarn fails to compile blake2

```
[user@host substrate-telemetry]$ nvm install stable
Downloading and installing node v12.10.0...
Downloading https://nodejs.org/dist/v12.10.0/node-v12.10.0-linux-x64.tar.xz...
########################################################################################################################################################################################################## 100.0%
Computing checksum with sha256sum
Checksums matched!
Now using node v12.10.0 (npm v6.10.3)
Creating default alias: default -> stable (-> v12.10.0)
[user@host substrate-telemetry]$ yarn
yarn install v1.17.3
warning package-lock.json found. Your project contains lock files generated by tools other than Yarn. It is advised not to mix package managers in order to avoid resolution inconsistencies caused by unsynchronized lock files. To clear this warning, remove package-lock.json.
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
info fsevents@1.2.6: The platform "linux" is incompatible with this module.
info "fsevents@1.2.6" is an optional dependency and failed compatibility check. Excluding it from installation.
[4/5] Linking dependencies...
warning "workspace-aggregator-22a41f02-bdec-4a46-bfa6-7118392a6dfe > @dotstats/frontend > react-svg@4.1.8" has unmet peer dependency "prop-types@^15.6.2".
warning "workspace-aggregator-22a41f02-bdec-4a46-bfa6-7118392a6dfe > @dotstats/frontend > babel-jest@23.6.0" has unmet peer dependency "babel-core@^6.0.0 || ^7.0.0-0".
warning "workspace-aggregator-22a41f02-bdec-4a46-bfa6-7118392a6dfe > @dotstats/frontend > react-scripts-ts > babel-jest@22.4.4" has unmet peer dependency "babel-core@^6.0.0 || ^7.0.0-0".
warning "workspace-aggregator-22a41f02-bdec-4a46-bfa6-7118392a6dfe > @dotstats/frontend > react-scripts-ts > babel-loader@7.1.5" has unmet peer dependency "babel-core@6".
warning "workspace-aggregator-22a41f02-bdec-4a46-bfa6-7118392a6dfe > @dotstats/frontend > react-scripts-ts > babel-preset-react-app@3.1.2" has unmet peer dependency "babel-runtime@^6.23.0".
[5/5] Building fresh packages...
[1/2] ⠂ blake2js
error /home/user/substrate-telemetry/node_modules/blake2js: Command failed.
Exit code: 1
Command: node-gyp rebuild
Arguments: 
Directory: /home/user/substrate-telemetry/node_modules/blake2js
Output:
gyp info it worked if it ends with ok
gyp info using node-gyp@5.0.3
gyp info using node@12.10.0 | linux | x64
gyp info find Python using Python version 2.7.16 found at "/usr/bin/python2"
gyp info spawn /usr/bin/python2
gyp info spawn args [
gyp info spawn args   '/home/user/.nvm/versions/node/v12.10.0/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/home/user/substrate-telemetry/node_modules/blake2js/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/home/user/.nvm/versions/node/v12.10.0/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/home/user/.cache/node-gyp/12.10.0/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/home/user/.cache/node-gyp/12.10.0',
gyp info spawn args   '-Dnode_gyp_dir=/home/user/.nvm/versions/node/v12.10.0/lib/node_modules/npm/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/home/user/.cache/node-gyp/12.10.0/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/home/user/substrate-telemetry/node_modules/blake2js',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.'
gyp info spawn args ]
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
make: Entering directory '/home/user/substrate-telemetry/node_modules/blake2js/build'
  CC(target) Release/obj.target/addon/lib/BLAKE2/sse/blake2s.o
../lib/BLAKE2/sse/blake2s.c:48:22: warning: ‘blake2s_sigma’ defined but not used [-Wunused-const-variable=]
   48 | static const uint8_t blake2s_sigma[10][16] =
      |                      ^~~~~~~~~~~~~
  CC(target) Release/obj.target/addon/lib/BLAKE2/sse/blake2sp.o
  CC(target) Release/obj.target/addon/lib/BLAKE2/sse/blake2b.o
../lib/BLAKE2/sse/blake2b.c:52:22: warning: ‘blake2b_sigma’ defined but not used [-Wunused-const-variable=]
   52 | static const uint8_t blake2b_sigma[12][16] =
      |                      ^~~~~~~~~~~~~
  CC(target) Release/obj.target/addon/lib/BLAKE2/sse/blake2bp.o
  CXX(target) Release/obj.target/addon/src/addon.o
In file included from ../../nan/nan_converters.h:67,
                 from ../../nan/nan.h:222,
                 from ../src/addon.cc:1:
../../nan/nan_converters_43_inl.h: In static member function ‘static Nan::imp::ToFactoryBase<v8::Boolean>::return_t Nan::imp::ToFactory<v8::Boolean>::convert(v8::Local<v8::Value>)’:
../../nan/nan_converters_43_inl.h:18:51: warning: ‘v8::MaybeLocal<v8::Boolean> v8::Value::ToBoolean(v8::Local<v8::Context>) const’ is deprecated: ToBoolean can never throw. Use Local version. [-Wdeprecated-declarations]
   18 |       val->To ## TYPE(isolate->GetCurrentContext())                            \
      |                                                   ^
../../nan/nan_converters_43_inl.h:22:1: note: in expansion of macro ‘X’
   22 | X(Boolean)
      | ^
In file included from /home/user/.cache/node-gyp/12.10.0/include/node/v8-internal.h:14,
                 from /home/user/.cache/node-gyp/12.10.0/include/node/v8.h:25,
                 from /home/user/.cache/node-gyp/12.10.0/include/node/node.h:63,
                 from ../../nan/nan.h:53,
                 from ../src/addon.cc:1:
/home/user/.cache/node-gyp/12.10.0/include/node/v8.h:2537:59: note: declared here
 2537 |                 V8_WARN_UNUSED_RESULT MaybeLocal<Boolean> ToBoolean(
      |                                                           ^~~~~~~~~
/home/user/.cache/node-gyp/12.10.0/include/node/v8config.h:311:3: note: in definition of macro ‘V8_DEPRECATED’
  311 |   declarator __attribute__((deprecated(message)))
      |   ^~~~~~~~~~
In file included from ../../nan/nan_converters.h:67,
                 from ../../nan/nan.h:222,
                 from ../src/addon.cc:1:
../../nan/nan_converters_43_inl.h: In static member function ‘static Nan::imp::ValueFactoryBase<bool>::return_t Nan::imp::ToFactory<bool>::convert(v8::Local<v8::Value>)’:
../../nan/nan_converters_43_inl.h:37:57: warning: ‘v8::Maybe<bool> v8::Value::BooleanValue(v8::Local<v8::Context>) const’ is deprecated: BooleanValue can never throw. Use Isolate version. [-Wdeprecated-declarations]
   37 |   return val->NAME ## Value(isolate->GetCurrentContext());                     \
      |                                                         ^
../../nan/nan_converters_43_inl.h:40:1: note: in expansion of macro ‘X’
   40 | X(bool, Boolean)
      | ^
In file included from /home/user/.cache/node-gyp/12.10.0/include/node/v8-internal.h:14,
                 from /home/user/.cache/node-gyp/12.10.0/include/node/v8.h:25,
                 from /home/user/.cache/node-gyp/12.10.0/include/node/node.h:63,
                 from ../../nan/nan.h:53,
                 from ../src/addon.cc:1:
/home/user/.cache/node-gyp/12.10.0/include/node/v8.h:2575:51: note: declared here
 2575 |                 V8_WARN_UNUSED_RESULT Maybe<bool> BooleanValue(
      |                                                   ^~~~~~~~~~~~
/home/user/.cache/node-gyp/12.10.0/include/node/v8config.h:311:3: note: in definition of macro ‘V8_DEPRECATED’
  311 |   declarator __attribute__((deprecated(message)))
      |   ^~~~~~~~~~
In file included from ../../nan/nan_new.h:189,
                 from ../../nan/nan.h:223,
                 from ../src/addon.cc:1:
../../nan/nan_implementation_12_inl.h: In static member function ‘static Nan::imp::FactoryBase<v8::StringObject>::return_t Nan::imp::Factory<v8::StringObject>::New(v8::Local<v8::String>)’:
../../nan/nan_implementation_12_inl.h:356:37: error: no matching function for call to ‘v8::StringObject::New(v8::Local<v8::String>&)’
  356 |   return v8::StringObject::New(value).As<v8::StringObject>();
      |                                     ^
In file included from /home/user/.cache/node-gyp/12.10.0/include/node/node.h:63,
                 from ../../nan/nan.h:53,
                 from ../src/addon.cc:1:
/home/user/.cache/node-gyp/12.10.0/include/node/v8.h:5394:23: note: candidate: ‘static v8::Local<v8::Value> v8::StringObject::New(v8::Isolate*, v8::Local<v8::String>)’
 5394 |   static Local<Value> New(Isolate* isolate, Local<String> value);
      |                       ^~~
/home/user/.cache/node-gyp/12.10.0/include/node/v8.h:5394:23: note:   candidate expects 2 arguments, 1 provided
In file included from ../../nan/nan_new.h:189,
                 from ../../nan/nan.h:223,
                 from ../src/addon.cc:1:
../../nan/nan_implementation_12_inl.h:356:58: error: expected primary-expression before ‘>’ token
  356 |   return v8::StringObject::New(value).As<v8::StringObject>();
      |                                                          ^
../../nan/nan_implementation_12_inl.h:356:60: error: expected primary-expression before ‘)’ token
  356 |   return v8::StringObject::New(value).As<v8::StringObject>();
      |                                                            ^
In file included from ../src/addon.cc:1:
../../nan/nan.h: In function ‘void Nan::AsyncQueueWorker(Nan::AsyncWorker*)’:
../../nan/nan.h:2232:62: warning: cast between incompatible function types from ‘void (*)(uv_work_t*)’ {aka ‘void (*)(uv_work_s*)’} to ‘uv_after_work_cb’ {aka ‘void (*)(uv_work_s*, int)’} [-Wcast-function-type]
 2232 |     , reinterpret_cast<uv_after_work_cb>(AsyncExecuteComplete)
      |                                                              ^
In file included from /usr/include/c++/9.1.0/cassert:44,
                 from /home/user/.cache/node-gyp/12.10.0/include/node/node_object_wrap.h:26,
                 from ../../nan/nan.h:55,
                 from ../src/addon.cc:1:
../../nan/nan_object_wrap.h: In destructor ‘virtual Nan::ObjectWrap::~ObjectWrap()’:
../../nan/nan_object_wrap.h:24:25: error: ‘class Nan::Persistent<v8::Object>’ has no member named ‘IsNearDeath’
   24 |     assert(persistent().IsNearDeath());
      |                         ^~~~~~~~~~~
../../nan/nan_object_wrap.h: In static member function ‘static void Nan::ObjectWrap::WeakCallback(const v8::WeakCallbackInfo<Nan::ObjectWrap>&)’:
../../nan/nan_object_wrap.h:127:26: error: ‘class Nan::Persistent<v8::Object>’ has no member named ‘IsNearDeath’
  127 |     assert(wrap->handle_.IsNearDeath());
      |                          ^~~~~~~~~~~
../src/addon.cc: In function ‘void B2JS::Init(v8::Local<v8::Object>)’:
../src/addon.cc:12:57: warning: ‘bool v8::Object::Set(v8::Local<v8::Value>, v8::Local<v8::Value>)’ is deprecated: Use maybe version [-Wdeprecated-declarations]
   12 |  exports->Set(New("blake2s").ToLocalChecked(), blake2s());
      |                                                         ^
In file included from /home/user/.cache/node-gyp/12.10.0/include/node/v8-internal.h:14,
                 from /home/user/.cache/node-gyp/12.10.0/include/node/v8.h:25,
                 from /home/user/.cache/node-gyp/12.10.0/include/node/node.h:63,
                 from ../../nan/nan.h:53,
                 from ../src/addon.cc:1:
/home/user/.cache/node-gyp/12.10.0/include/node/v8.h:3372:22: note: declared here
 3372 |                 bool Set(Local<Value> key, Local<Value> value));
      |                      ^~~
/home/user/.cache/node-gyp/12.10.0/include/node/v8config.h:311:3: note: in definition of macro ‘V8_DEPRECATED’
  311 |   declarator __attribute__((deprecated(message)))
      |   ^~~~~~~~~~
../src/addon.cc:13:59: warning: ‘bool v8::Object::Set(v8::Local<v8::Value>, v8::Local<v8::Value>)’ is deprecated: Use maybe version [-Wdeprecated-declarations]
   13 |  exports->Set(New("blake2sp").ToLocalChecked(), blake2sp());
      |                                                           ^
In file included from /home/user/.cache/node-gyp/12.10.0/include/node/v8-internal.h:14,
                 from /home/user/.cache/node-gyp/12.10.0/include/node/v8.h:25,
                 from /home/user/.cache/node-gyp/12.10.0/include/node/node.h:63,
                 from ../../nan/nan.h:53,
                 from ../src/addon.cc:1:
/home/user/.cache/node-gyp/12.10.0/include/node/v8.h:3372:22: note: declared here
 3372 |                 bool Set(Local<Value> key, Local<Value> value));
      |                      ^~~
/home/user/.cache/node-gyp/12.10.0/include/node/v8config.h:311:3: note: in definition of macro ‘V8_DEPRECATED’
  311 |   declarator __attribute__((deprecated(message)))
      |   ^~~~~~~~~~
../src/addon.cc:14:57: warning: ‘bool v8::Object::Set(v8::Local<v8::Value>, v8::Local<v8::Value>)’ is deprecated: Use maybe version [-Wdeprecated-declarations]
   14 |  exports->Set(New("blake2b").ToLocalChecked(), blake2b());
      |                                                         ^
In file included from /home/user/.cache/node-gyp/12.10.0/include/node/v8-internal.h:14,
                 from /home/user/.cache/node-gyp/12.10.0/include/node/v8.h:25,
                 from /home/user/.cache/node-gyp/12.10.0/include/node/node.h:63,
                 from ../../nan/nan.h:53,
                 from ../src/addon.cc:1:
/home/user/.cache/node-gyp/12.10.0/include/node/v8.h:3372:22: note: declared here
 3372 |                 bool Set(Local<Value> key, Local<Value> value));
      |                      ^~~
/home/user/.cache/node-gyp/12.10.0/include/node/v8config.h:311:3: note: in definition of macro ‘V8_DEPRECATED’
  311 |   declarator __attribute__((deprecated(message)))
      |   ^~~~~~~~~~
../src/addon.cc:15:59: warning: ‘bool v8::Object::Set(v8::Local<v8::Value>, v8::Local<v8::Value>)’ is deprecated: Use maybe version [-Wdeprecated-declarations]
   15 |  exports->Set(New("blake2bp").ToLocalChecked(), blake2bp());
      |                                                           ^
In file included from /home/user/.cache/node-gyp/12.10.0/include/node/v8-internal.h:14,
                 from /home/user/.cache/node-gyp/12.10.0/include/node/v8.h:25,
                 from /home/user/.cache/node-gyp/12.10.0/include/node/node.h:63,
                 from ../../nan/nan.h:53,
                 from ../src/addon.cc:1:
/home/user/.cache/node-gyp/12.10.0/include/node/v8.h:3372:22: note: declared here
 3372 |                 bool Set(Local<Value> key, Local<Value> value));
      |                      ^~~
/home/user/.cache/node-gyp/12.10.0/include/node/v8config.h:311:3: note: in definition of macro ‘V8_DEPRECATED’
  311 |   declarator __attribute__((deprecated(message)))
      |   ^~~~~~~~~~
In file included from ../../nan/nan.h:53,
                 from ../src/addon.cc:1:
../src/addon.cc: At global scope:
/home/user/.cache/node-gyp/12.10.0/include/node/node.h:560:43: warning: cast between incompatible function types from ‘void (*)(v8::Local<v8::Object>)’ to ‘node::addon_register_func’ {aka ‘void (*)(v8::Local<v8::Object>, v8::Local<v8::Value>, void*)’} [-Wcast-function-type]
  560 |       (node::addon_register_func) (regfunc),                          \
      |                                           ^
/home/user/.cache/node-gyp/12.10.0/include/node/node.h:594:3: note: in expansion of macro ‘NODE_MODULE_X’
  594 |   NODE_MODULE_X(modname, regfunc, NULL, 0)  // NOLINT (readability/null_usage)
      |   ^~~~~~~~~~~~~
../src/addon.cc:18:1: note: in expansion of macro ‘NODE_MODULE’
   18 | NODE_MODULE(addon, Init)
      | ^~~~~~~~~~~
In file included from /home/user/.cache/node-gyp/12.10.0/include/node/node.h:63,
                 from ../../nan/nan.h:53,
                 from ../src/addon.cc:1:
/home/user/.cache/node-gyp/12.10.0/include/node/v8.h: In instantiation of ‘void v8::PersistentBase<T>::SetWeak(P*, typename v8::WeakCallbackInfo<P>::Callback, v8::WeakCallbackType) [with P = node::ObjectWrap; T = v8::Object; typename v8::WeakCallbackInfo<P>::Callback = void (*)(const v8::WeakCallbackInfo<node::ObjectWrap>&)]’:
/home/user/.cache/node-gyp/12.10.0/include/node/node_object_wrap.h:84:78:   required from here
/home/user/.cache/node-gyp/12.10.0/include/node/v8.h:9893:16: warning: cast between incompatible function types from ‘v8::WeakCallbackInfo<node::ObjectWrap>::Callback’ {aka ‘void (*)(const v8::WeakCallbackInfo<node::ObjectWrap>&)’} to ‘Callback’ {aka ‘void (*)(const v8::WeakCallbackInfo<void>&)’} [-Wcast-function-type]
 9893 |                reinterpret_cast<Callback>(callback), type);
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/user/.cache/node-gyp/12.10.0/include/node/v8.h: In instantiation of ‘void v8::PersistentBase<T>::SetWeak(P*, typename v8::WeakCallbackInfo<P>::Callback, v8::WeakCallbackType) [with P = Nan::ObjectWrap; T = v8::Object; typename v8::WeakCallbackInfo<P>::Callback = void (*)(const v8::WeakCallbackInfo<Nan::ObjectWrap>&)]’:
../../nan/nan_object_wrap.h:66:61:   required from here
/home/user/.cache/node-gyp/12.10.0/include/node/v8.h:9893:16: warning: cast between incompatible function types from ‘v8::WeakCallbackInfo<Nan::ObjectWrap>::Callback’ {aka ‘void (*)(const v8::WeakCallbackInfo<Nan::ObjectWrap>&)’} to ‘Callback’ {aka ‘void (*)(const v8::WeakCallbackInfo<void>&)’} [-Wcast-function-type]
make: *** [addon.target.mk:118: Release/obj.target/addon/src/addon.o] Error 1
make: Leaving directory '/home/user/substrate-telemetry/node_modules/blake2js/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/home/user/.nvm/versions/node/v12.10.0/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:196:23)
gyp ERR! stack     at ChildProcess.emit (events.js:209:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:272:12)
gyp ERR! System Linux 5.2.11-arch1-1-ARCH
gyp ERR! command "/home/user/.nvm/versions/node/v12.10.0/bin/node" "/home/user/.nvm/versions/node/v12.10.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/user/substrate-telemetry/node_modules/blake2js
gyp ERR! node -v v12.10.0
gyp ERR! node-gyp -v v5.0.3
gyp ERR! not ok
```

--- a/backend/src/chain.rs
+++ b/backend/src/chain.rs
@@ -257,6 +257,10 @@ impl Handler<UpdateNode> for Chain {
                         node.set_network_state(raw);
                     }
                 }
+                Details::AfgAuthoritySet(authority) => {
+                    node.set_validator_address(authority.authority_id);
+                    return;
+                }
                 _ => (),
             }
 

--- a/backend/src/chain.rs
+++ b/backend/src/chain.rs
@@ -204,7 +204,7 @@ impl Handler<UpdateNode> for Chain {
         let UpdateNode { nid, msg, raw } = msg;
 
         if let Some(block) = msg.details.best_block() {
-            let mut propagation_time = 0;
+            let mut propagation_time = None;
             let now = now();
 
             self.update_stale_nodes(now);
@@ -224,7 +224,7 @@ impl Handler<UpdateNode> for Chain {
                 self.serializer.push(feed::BestBlock(self.best.height, now, self.average_block_time));
             } else if block.height == self.best.height {
                 if let Some(timestamp) = self.timestamp {
-                    propagation_time = now - timestamp;
+                    propagation_time = Some(now - timestamp);
                 }
             }
 

--- a/backend/src/node.rs
+++ b/backend/src/node.rs
@@ -55,7 +55,7 @@ impl Node {
                 block: Block::zero(),
                 block_timestamp: now(),
                 block_time: 0,
-                propagation_time: 0,
+                propagation_time: None,
             },
             finalized: Block::zero(),
             throttle: 0,
@@ -111,7 +111,7 @@ impl Node {
         &self.best
     }
 
-    pub fn update_block(&mut self, block: Block, timestamp: u64, propagation_time: u64) -> Option<&BlockDetails> {
+    pub fn update_block(&mut self, block: Block, timestamp: u64, propagation_time: Option<u64>) -> Option<&BlockDetails> {
         if block.height > self.best.block.height {
             self.stale = false;
             self.best.block = block;

--- a/backend/src/node.rs
+++ b/backend/src/node.rs
@@ -182,6 +182,10 @@ impl Node {
         self.stale
     }
 
+    pub fn set_validator_address(&mut self, addr: Box<str>) {
+        self.details.validator = Some(addr);
+    }
+
     pub fn set_network_state(&mut self, state: Bytes) {
         self.network_state = Some(state);
     }

--- a/backend/src/node/message.rs
+++ b/backend/src/node/message.rs
@@ -45,7 +45,7 @@ pub enum Details {
     #[serde(rename = "afg.received_commit")]
     AfgReceivedCommit(IgnoredAny),
     #[serde(rename = "afg.authority_set")]
-    AfgAuthoritySet(IgnoredAny),
+    AfgAuthoritySet(AfgAuthoritySet),
     #[serde(rename = "aura.pre_sealed_block")]
     AuraPreSealedBlock(IgnoredAny),
     #[serde(rename = "prepared_block_for_proposing")]
@@ -79,6 +79,11 @@ pub struct Finalized {
     #[serde(rename = "best")]
     pub hash: BlockHash,
     pub height: Box<str>,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct AfgAuthoritySet {
+    pub authority_id: Box<str>,
 }
 
 impl Block {

--- a/backend/src/node/message.rs
+++ b/backend/src/node/message.rs
@@ -46,6 +46,8 @@ pub enum Details {
     AfgReceivedCommit(IgnoredAny),
     #[serde(rename = "afg.authority_set")]
     AfgAuthoritySet(AfgAuthoritySet),
+    #[serde(rename = "afg.finalized_blocks_up_to")]
+    AfgFinalizedBlocksUpTo(IgnoredAny),
     #[serde(rename = "aura.pre_sealed_block")]
     AuraPreSealedBlock(IgnoredAny),
     #[serde(rename = "prepared_block_for_proposing")]

--- a/backend/src/types.rs
+++ b/backend/src/types.rs
@@ -12,6 +12,7 @@ pub struct NodeDetails {
     pub implementation: Box<str>,
     pub version: Box<str>,
     pub validator: Option<Box<str>>,
+    pub network_id: Option<Box<str>>,
 }
 
 #[derive(Deserialize, Debug, Clone, Copy, PartialEq, Eq)]
@@ -54,7 +55,7 @@ impl Serialize for NodeDetails {
         tup.serialize_element(&self.implementation)?;
         tup.serialize_element(&self.version)?;
         tup.serialize_element(&self.validator)?; // TODO Maybe<Address>
-        tup.serialize_element::<Option<usize>>(&None)?; // TODO Maybe<NetworkId>
+        tup.serialize_element(&self.network_id)?; // TODO Maybe<NetworkId>
         tup.serialize_element("")?; // TODO Address
         tup.end()
     }

--- a/backend/src/types.rs
+++ b/backend/src/types.rs
@@ -32,7 +32,7 @@ pub struct BlockDetails {
     pub block: Block,
     pub block_time: u64,
     pub block_timestamp: u64,
-    pub propagation_time: u64,
+    pub propagation_time: Option<u64>,
 }
 
 pub type NodeHardware<'a> = (&'a [f32], &'a [f32], &'a [f64], &'a [f64], &'a [f64]);

--- a/backend/src/types.rs
+++ b/backend/src/types.rs
@@ -11,6 +11,7 @@ pub struct NodeDetails {
     pub name: Box<str>,
     pub implementation: Box<str>,
     pub version: Box<str>,
+    pub validator: Option<Box<str>>,
 }
 
 #[derive(Deserialize, Debug, Clone, Copy, PartialEq, Eq)]
@@ -52,7 +53,7 @@ impl Serialize for NodeDetails {
         tup.serialize_element(&self.name)?;
         tup.serialize_element(&self.implementation)?;
         tup.serialize_element(&self.version)?;
-        tup.serialize_element::<Option<String>>(&None)?; // TODO Maybe<Address>
+        tup.serialize_element(&self.validator)?; // TODO Maybe<Address>
         tup.serialize_element::<Option<usize>>(&None)?; // TODO Maybe<NetworkId>
         tup.serialize_element("")?; // TODO Address
         tup.end()

--- a/packages/frontend/package.json
+++ b/packages/frontend/package.json
@@ -21,13 +21,14 @@
   "dependencies": {
     "@fnando/sparkline": "maciejhirsz/sparkline",
     "@types/react-measure": "^2.0.5",
-    "polkadot-identicon": "^1.1.45",
+    "@polkadot/util-crypto": "^1.6.1",
     "react": "16.4.0",
     "react-dom": "16.4.0",
     "react-measure": "^2.3.0",
     "react-scripts-ts": "2.17.0",
     "react-svg": "^4.1.1",
-    "stable": "^0.1.8"
+    "stable": "^0.1.8",
+    "blakejs": "^1.1.0"
   },
   "scripts": {
     "start": "react-scripts-ts start",

--- a/packages/frontend/src/components/Consensus/ConsensusBlock.tsx
+++ b/packages/frontend/src/components/Consensus/ConsensusBlock.tsx
@@ -2,9 +2,8 @@ import * as React from 'react';
 
 import Measure, {BoundingRect, ContentRect} from 'react-measure';
 import { Types, Maybe } from '@dotstats/common';
-import Identicon from 'polkadot-identicon';
 
-import { Icon, Tooltip } from '../';
+import { Icon, Tooltip, PolkadotIcon } from '../';
 import Jdenticon from './Jdenticon';
 
 import checkIcon from '../../icons/check.svg';
@@ -183,7 +182,7 @@ export class ConsensusBlock extends React.Component<ConsensusBlock.Props, {}> {
   private getAuthorityContent(authority: Types.Authority): JSX.Element {
     return <div className="nodeContent" key={'authority_' + this.props.height + '_' + authority.Address}>
       <div className="nodeAddress" key={'authority_' + authority.Address}>
-        <Identicon account={authority.Address} size={this.props.compact ? 14 : 28} />
+        <PolkadotIcon account={authority.Address} size={this.props.compact ? 14 : 28} />
       </div>
     </div>;
   }

--- a/packages/frontend/src/components/List/Row.tsx
+++ b/packages/frontend/src/components/List/Row.tsx
@@ -1,11 +1,10 @@
 import * as React from 'react';
-import Identicon from 'polkadot-identicon';
 import { Types, Maybe, timestamp } from '@dotstats/common';
 import { formatNumber, getHashData, milliOrSecond, secondsWithPrecision } from '../../utils';
 import { State as AppState, Node } from '../../state';
 import { PersistentSet } from '../../persist';
 import { Truncate } from './';
-import { Ago, Icon, Tooltip, Sparkline } from '../';
+import { Ago, Icon, Tooltip, Sparkline, PolkadotIcon } from '../';
 
 import nodeIcon from '../../icons/server.svg';
 import nodeLocationIcon from '../../icons/location.svg';
@@ -150,7 +149,7 @@ export class Row extends React.Component<Row.Props, Row.State> {
       width: 16,
       setting: 'validator',
       render: ({ validator }) => {
-        return validator ? <Tooltip text={validator} copy={true}><span className="Row-validator"><Identicon account={validator} size={16} /></span></Tooltip> : '-';
+        return validator ? <Tooltip text={validator} copy={true}><span className="Row-validator"><PolkadotIcon account={validator} size={16} /></span></Tooltip> : '-';
       }
     },
     {

--- a/packages/frontend/src/components/Map/Location.tsx
+++ b/packages/frontend/src/components/Map/Location.tsx
@@ -1,7 +1,6 @@
 import * as React from 'react';
-import Identicon from 'polkadot-identicon';
 import { formatNumber, trimHash, milliOrSecond, secondsWithPrecision } from '../../utils';
-import { Ago, Icon } from '../';
+import { Ago, Icon, PolkadotIcon } from '../';
 import { Node } from '../../state';
 
 import nodeIcon from '../../icons/server.svg';
@@ -92,7 +91,7 @@ export class Location extends React.Component<Location.Props, Location.State> {
           <td><Icon src={nodeValidatorIcon} alt="Node" /></td>
           <td colSpan={5}>
             {trimHash(validator, 30)}
-            <span className="Location-validator"><Identicon account={validator} size={16} /></span>
+            <span className="Location-validator"><PolkadotIcon account={validator} size={16} /></span>
           </td>
         </tr>
       );

--- a/packages/frontend/src/components/PolkadotIcon.tsx
+++ b/packages/frontend/src/components/PolkadotIcon.tsx
@@ -1,0 +1,187 @@
+// Copyright 2018 Paritytech via paritytech/oo7/polkadot-identicon
+// Copyright 2018 @polkadot/ui-shared authors & contributors
+// This software may be modified and distributed under the terms
+// of the Apache-2.0 license. See the LICENSE file for details.
+
+// This has been converted from the original version that can be found at
+//
+// https://github.com/paritytech/oo7/blob/251ba2b7c45503b68eab4320c270b5afa9bccb60/packages/polkadot-identicon/src/index.jsx
+import * as React from 'react';
+import { blake2AsU8a, decodeAddress } from '@polkadot/util-crypto';
+
+interface Circle {
+  cx: number;
+  cy: number;
+  fill: string;
+  r: number;
+}
+
+interface Scheme {
+  freq: number;
+  colors: number[];
+}
+
+const blake2 = (value: Uint8Array): Uint8Array =>
+  blake2AsU8a(value, 512);
+
+const S = 64;
+const C = S / 2;
+const Z = S / 64 * 5;
+const ZERO = blake2(new Uint8Array(32));
+
+const SCHEMA: Scheme[] = [
+  // target
+  { freq: 1, colors: [0, 28, 0, 0, 28, 0, 0, 28, 0, 0, 28, 0, 0, 28, 0, 0, 28, 0, 1] },
+  // cube
+  { freq: 20, colors: [0, 1, 3, 2, 4, 3, 0, 1, 3, 2, 4, 3, 0, 1, 3, 2, 4, 3, 5] },
+  // quazar
+  { freq: 16, colors: [1, 2, 3, 1, 2, 4, 5, 5, 4, 1, 2, 3, 1, 2, 4, 5, 5, 4, 0] },
+  // flower
+  { freq: 32, colors: [0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 3] },
+  // cyclic
+  { freq: 32, colors: [0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 6] },
+  // vmirror
+  { freq: 128, colors: [0, 1, 2, 3, 4, 5, 3, 4, 2, 0, 1, 6, 7, 8, 9, 7, 8, 6, 10] },
+  // hmirror
+  { freq: 128, colors: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 8, 6, 7, 5, 3, 4, 2, 11] },
+];
+
+const OUTER_CIRCLE: Circle = {
+  cx: C,
+  cy: C,
+  r: C,
+  fill: '#eee'
+};
+
+function getRotation (isSixPoint: boolean): { r: number; ro2: number; r3o4: number; ro4: number; rroot3o2: number; rroot3o4: number } {
+  const r = isSixPoint
+    ? (C / 8 * 5)
+    : (C / 4 * 3);
+  const rroot3o2 = r * Math.sqrt(3) / 2;
+  const ro2 = r / 2;
+  const rroot3o4 = r * Math.sqrt(3) / 4;
+  const ro4 = r / 4;
+  const r3o4 = r * 3 / 4;
+
+  return { r, ro2, r3o4, ro4, rroot3o2, rroot3o4 };
+}
+
+function getCircleXY (isSixPoint: boolean): Array<[number, number]> {
+  const { r, ro2, r3o4, ro4, rroot3o2, rroot3o4 } = getRotation(isSixPoint);
+
+  return [
+    [C, C - r],
+    [C, C - ro2],
+    [C - rroot3o4, C - r3o4],
+    [C - rroot3o2, C - ro2],
+    [C - rroot3o4, C - ro4],
+    [C - rroot3o2, C],
+    [C - rroot3o2, C + ro2],
+    [C - rroot3o4, C + ro4],
+    [C - rroot3o4, C + r3o4],
+    [C, C + r],
+    [C, C + ro2],
+    [C + rroot3o4, C + r3o4],
+    [C + rroot3o2, C + ro2],
+    [C + rroot3o4, C + ro4],
+    [C + rroot3o2, C],
+    [C + rroot3o2, C - ro2],
+    [C + rroot3o4, C - ro4],
+    [C + rroot3o4, C - r3o4],
+    [C, C]
+  ];
+}
+
+function findScheme (d: number): Scheme {
+  let sum = 0;
+  const schema = SCHEMA.find((s): boolean => {
+    sum += s.freq;
+
+    return d < sum;
+  });
+
+  if (!schema) {
+    throw new Error('Unable to find schema');
+  }
+
+  return schema;
+}
+
+function addressToId (address: string): Uint8Array {
+  return blake2(decodeAddress(address)).map((x, i): number => (x + 256 - ZERO[i]) % 256);
+}
+
+function getColors (address: string): string[] {
+  const total = SCHEMA.map((s): number => s.freq).reduce((a, b): number => a + b);
+  const id = addressToId(address);
+  const d = Math.floor((id[30] + id[31] * 256) % total);
+  const rot = (id[28] % 6) * 3;
+  const sat = (Math.floor(id[29] * 70 / 256 + 26) % 80) + 30;
+  const scheme = findScheme(d);
+  const palette = Array.from(id).map((x, i): string => {
+    const b = (x + i % 28 * 58) % 256;
+
+    if (b === 0) {
+      return '#444';
+    } else if (b === 255) {
+      return 'transparent';
+    }
+
+    const h = Math.floor(b % 64 * 360 / 64);
+    const l = [53, 15, 35, 75][Math.floor(b / 64)];
+
+    return `hsl(${h}, ${sat}%, ${l}%)`;
+  });
+
+  return scheme.colors.map((_, i): string =>
+    palette[scheme.colors[i < 18 ? (i + rot) % 18 : 18]]
+  );
+}
+
+/**
+ * @description Generate a array of the circles that make up an indenticon
+ */
+export default function generate (address: string, isSixPoint = false): Circle[] {
+  const colors = getColors(address);
+
+  return [OUTER_CIRCLE].concat(
+    getCircleXY(isSixPoint).map(([cx, cy], index): Circle => ({
+      cx, cy, r: Z, fill: colors[index]
+    }))
+  );
+}
+
+export namespace PolkadotIcon {
+  export interface Props {
+    account: string;
+    size: number;
+  }
+}
+
+export class PolkadotIcon extends React.Component<PolkadotIcon.Props, {}> {
+  public render (): React.ReactNode {
+    const { account, size } = this.props;
+
+    return (
+      <svg
+        width={size}
+        height={size}
+        viewBox='0 0 64 64'
+      >
+        {generate(account, false).map(this.renderCircle)}
+      </svg>
+    );
+  }
+
+  private renderCircle = ({ cx, cy, r, fill }: Circle, key: number): React.ReactNode => {
+    return (
+      <circle
+        key={key}
+        cx={cx}
+        cy={cy}
+        r={r}
+        fill={fill}
+      />
+    );
+  }
+}

--- a/packages/frontend/src/components/index.ts
+++ b/packages/frontend/src/components/index.ts
@@ -12,3 +12,4 @@ export * from './OfflineIndicator';
 export * from './Sparkline';
 export * from './Tooltip';
 export * from './Filter';
+export * from './PolkadotIcon';

